lib/Sema/TypeCheckType.cpp

```diff
@@ -5664,11 +5665,26 @@ TypeResolver::resolveExistentialType(ExistentialTypeRepr *repr,
 
   //TO-DO: generalize this and emit the same erorr for some P?
   if (!constraintType->isConstraintType()) {
+    auto wrapped = constraintType->getOptionalObjectType();
+
+    auto isWrappedExistential = (wrapped && (wrapped->is<ExistentialType>() || wrapped->is<ExistentialMetatypeType>())) ;
+    auto isExistentialInsideTuple = options.is(TypeResolverContext::TupleElement);
+    auto isImplicitlyUnwrapped =  isa<ImplicitlyUnwrappedOptionalTypeRepr>(repr->getConstraint()) ;
+    // Handles cases like `let _: any P!` by suggesting `let _: (any P)!
+    // while making sure that it's not an existential wrapped in a tuple like `let _: (Int, any P!)`
+    // so we won't suggest `let _: (Int, (any P)!)` as it's illegal.
+    if (isImplicitlyUnwrapped && !isExistentialInsideTuple && isWrappedExistential)  {
+      diagnose(repr->getLoc(),
+               diag::incorrect_force_unwrapped_any,
+               wrapped->getMetatypeInstanceType().getString())
+          .fixItReplace(repr->getSourceRange(), "(" + wrapped.getString() + ")!");
```

Comment:
I see that the approach you have taken is to skip diagnosing `!` in `resolveImplicitlyUnwrappedOptionalType`  when the context is an explicit existential constraint, and handle this case in `resolveExistentialType` instead, which I reckon is a reasonable idea. Here are some possible directions for improvement:
* The fundamental problem with `(Int, any P!)` is that the language forbids us to use a `!` here even if you write `(any P)!`. I believe it is important to clearly communicate this rule. In this case, `incorrect_optional_any` is a bit confusing and sounds like the compiler is not paying enough attention to the intention of the user. What if we diagnose the `!` exactly like `resolveImplicitlyUnwrappedOptionalType` does, but with a fix-it that takes the existential into account?
* Tuple element types are just one possible nested type position where `!` is illegal. It could also be `G<any P!>`, `(any P!) -> Void`, `(any P!)?`, etc. For a comprehensive check, we could extract the relevant part of `resolveImplicitlyUnwrappedOptionalType` into a function called `shouldDiagnoseImplicitlyUnwrappedOptionalType`.